### PR TITLE
Check tasks (plans) on the Dashboard

### PR DESCRIPTION
**Pull request by matiasbavera**
## What's new
pre-requisite https://github.com/osrf/romi-dashboard/pull/153 [Merged]

fixes #

* Created a taskSummary windows on the omnipannel with a tree structure, 
* Subscribed to TaskSummary
* Created mock
* Added a pause button but it is not implemented yet.
* Add unit tests
* Add e2e tests

image

## Self-checks

- [X] I'm familiar with and follow this [ Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [X] I added unit-tests for new components
- [X] I tried testing edge cases
- [X] I tested the behavior of the components that interact with the backend, with an e2e test

## Discussion

If we will keep the tree button group component I'll refactor the negotiation panel in another PR

### CODE REVIEW

**Comment by matiasbavera** [comment]
Converting to draft again, I Refactored the code, test are missing

File: packages/dashboard/src/components/task-summary-panel.tsx [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  const handleSelect = (event: React.ChangeEvent<{}>, nodeIds: string) => {
+    setSelected(nodeIds);
+  };
+
+  const handleResetTasks = () => {
+    setExpanded([]);
+    setSelected('');
+  };
+
+  const handleClearAllCurrTasks = () => {
```
Comment (koonpeng) [inline packages/dashboard/src/components/task-summary-panel.tsx:60]: Why would the user want to clear current tasks? From what I understand, this only clears the task from the list, and does not actually cancel the tasks.

File: packages/dashboard/src/managers/task-manager.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        this.emit('updated');
+      }),
+    );
+  }
+
+  stopAllSubscriptions(): void {
+    this._subscriptions.forEach((sub) => sub.unsubscribe());
+  }
+
+  static getActorFromStatus(status: string) {
```
Comment (koonpeng) [inline packages/dashboard/src/managers/task-manager.ts:26]: This looks like a hack, can you add a `TODO` and a issue (here and at `rmf_core`) to provide a reliable way to retrieve the actor from the message?

File: packages/dashboard/src/components/task-summary-panel.tsx [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        </TreeItem>
+      );
+    };
+
+    if (allTasks.length === 0) return;
+    let taskContentsTemp = Object.assign({}, taskContents);
+    allTasks.forEach((task) => {
+      taskContentsTemp[task.task_id] = renderTaskTreeItem(task);
+    });
+    setTaskContents(taskContentsTemp);
```
Comment (koonpeng) [inline packages/dashboard/src/components/task-summary-panel.tsx:143]: You should be able to use the functional form to avoid recursive dependencies and use of a temp object.
```tsx
setTaskContents((currentContents) => {
  return doStuff(currentContents);
});
```

File: packages/dashboard/src/components/task-summary-panel.tsx [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+  const handleRestoreTasks = () => {
+    // Assigning another reference
+    let taskContentsTemp = Object.assign({}, taskContents);
+    Object.keys(savedTasksContent.current).forEach((element) => {
+      if (!(element in taskContents)) {
+        taskContentsTemp[element] = savedTasksContent.current[element];
+      }
+    });
+    setTaskContents(taskContentsTemp);
```
Comment (koonpeng) [inline packages/dashboard/src/components/task-summary-panel.tsx:73]: You should be able to use the functional form to avoid recursive dependencies and use of a temp object.
```tsx
setTaskContents((currentContents) => {
  return doStuff(currentContents);
});
```

File: packages/dashboard/src/managers/task-manager.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    // Gets the name of the robot if it has any
+    // eslint-disable-next-line
+    return status.match(/\[[A-Za-z]([a-zA-Z0-9\/]){3,}\]+/gi);
+  }
+
+  static formatStatus(status: string) {
+    return status.split('|');
+  }
+
+  static getStateLabel(state: number): string {
```
Comment (koonpeng) [inline packages/dashboard/src/managers/task-manager.ts:36]: Better to keep presentation logic out of here, move it to the component instead.

**Review by koonpeng — CHANGES_REQUESTED** [review]

File: packages/dashboard/src/components/task-summary-panel.tsx (hunk 1 above)
Comment (matiasbavera) [reply packages/dashboard/src/components/task-summary-panel.tsx:60]: Yeah, it only clears the task from the list and does not cancel the task.

I want to implement the task pause in a FUP PR. 

I just thought that they will want to clear it when they have a lot of old finished tasks on the list.

File: packages/dashboard/src/managers/task-manager.ts (hunk 2 above)
Comment (matiasbavera) [reply packages/dashboard/src/managers/task-manager.ts:26]: Sure, sounds good! I'll create and issue and add a todo

Comment (matiasbavera) [reply packages/dashboard/src/managers/task-manager.ts:26]: https://github.com/osrf/rmf_core/issues/205

File: packages/dashboard/src/components/task-summary-panel.tsx (hunk 3 above)
Comment (matiasbavera) [reply packages/dashboard/src/components/task-summary-panel.tsx:143]: I looked into it and we cannot assign directly the stored values to `currentContent` because it is updating the taskContents too, so when we set the taskContent new value with `setTaskContents` it'll no re-render because it'll not detect any changes.
We can also force the re-render of the component

File: packages/dashboard/src/components/task-summary-panel.tsx (hunk 1 above)
Comment (koonpeng) [reply packages/dashboard/src/components/task-summary-panel.tsx:61]: Hmm, it's the lack of a backend database problem again. If we keep the previous tasks in memory it will be gone when we refresh the page, we can't persist the tasks without a db. I guess this is the best we can do now so that the user can see the result of the task. Can you add a TODO here?

File: packages/dashboard/src/components/task-summary-panel.tsx [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            root: classes.treeChildren,
+            expanded: classes.expanded,
+          }}
+          label={
+            <>
+              <Typography variant="body1" noWrap>
+                {task.state === RomiCore.TaskSummary.STATE_ACTIVE && (
+                  <IconButton
+                    // TODO: replace onClick with the pause plans logic.
+                    onClick={(e) => {
```
Comment (koonpeng) [inline packages/dashboard/src/components/task-summary-panel.tsx:131]: make it not clickable?

File: packages/dashboard/src/components/task-summary-panel.tsx [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    setTaskContents({});
+  };
+
+  const handleRestoreTasks = () => {
+    // Adding this because the test interpreter cannot find the reference to `savedTasksContent current`
+    // inside the setTaskContents
+    const storedTasks = savedTasksContent.current;
+    setTaskContents((currentContent) => {
+      // Assigning another reference.
+      let taskContentsTemp = Object.assign({}, currentContent);
```
Comment (koonpeng) [inline packages/dashboard/src/components/task-summary-panel.tsx:72]: use `const`. Also let's rename this to `newTaskContents` to better reflect its purpose.

File: packages/dashboard/src/components/task-summary-panel.tsx [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+      // We cannot assign directly the stored values to `currentContent` because it is updating
+      // the taskContents too, so when we set the taskContent new value with `setTaskContents`
+      // it'll no re-render because it'll not detect any changes.
+      Object.keys(storedTasks).forEach((element) => {
+        if (!(element in currentContent)) {
+          taskContentsTemp[element] = storedTasks[element];
+        }
+      });
+      console.log(taskContentsTemp);
```
Comment (koonpeng) [inline packages/dashboard/src/components/task-summary-panel.tsx:81]: leftover?

File: packages/dashboard/src/components/task-summary-panel.tsx [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+      // it'll no re-render because it'll not detect any changes.
+      allTasks.forEach((task) => {
+        taskContentsTemp[task.task_id] = renderTaskTreeItem(task);
+      });
+      return taskContentsTemp;
+    });
+
+    // Ignoring dependency taskContents, it'll enter in a infinite loop
+    // eslint-disable-next-line
+  }, [allTasks]);
```
Comment (koonpeng) [inline packages/dashboard/src/components/task-summary-panel.tsx:177]: eslint warning
```
React Hook React.useEffect has missing dependencies: 'classes.expanded', 'classes.labelContent', 'classes.taskActor', 'classes.treeChildren', and 'determineStyle'. Either include them or remove the dependency array.
```

File: packages/dashboard/src/components/task-summary-panel.tsx [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+      // the taskContents too, so when we set the taskContent new value with `setTaskContents`
+      // it'll no re-render because it'll not detect any changes.
+      allTasks.forEach((task) => {
+        taskContentsTemp[task.task_id] = renderTaskTreeItem(task);
+      });
+      return taskContentsTemp;
+    });
+
+    // Ignoring dependency taskContents, it'll enter in a infinite loop
+    // eslint-disable-next-line
```
Comment (koonpeng) [inline packages/dashboard/src/components/task-summary-panel.tsx:176]: This no longer holds a dependency on `taskContents` we can remove this.

File: packages/dashboard/src/components/task-summary-panel.tsx [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+          <TaskSummaryPanelItem task={task} key={task.task_id} />
+        </TreeItem>
+      );
+    };
+
+    if (allTasks.length === 0) return;
+
+    setTaskContents((currentContent) => {
+      // Assigning another reference.
+      let taskContentsTemp = Object.assign({}, currentContent);
```
Comment (koonpeng) [inline packages/dashboard/src/components/task-summary-panel.tsx:165]: use `const`. Also let's rename this to `newTaskContents` to better reflect its purpose.

**Review by koonpeng — CHANGES_REQUESTED** [review]

**Comment by koonpeng** [comment]
Peek 2020-11-06 09-37
I have a bug where the the color is "stucked" in the hover state after clicking the task.

File: packages/dashboard/src/components/task-summary-panel.tsx (hunk 1 above)
Comment (matiasbavera) [reply packages/dashboard/src/components/task-summary-panel.tsx:61]: Sure thing!

**Comment by matiasbavera** [comment]
> I have a bug where the the color is "stucked" in the hover state after clicking the task.

You're right, I'll take a look. Thanks

File: packages/dashboard/src/components/tree-button-group.tsx [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+          id="restore-button"
+          disabled={disableRestore}
+          onClick={() => handleRestoreClick && handleRestoreClick()}
+        >
+          <RestoreFromTrashIcon />
+          Restore
+        </Button>
+      </ButtonGroup>
+    </div>
+  );
```
Comment (matiasbavera) [inline packages/dashboard/src/components/tree-button-group.tsx:56]: Should I move this to `react-component` package on this PR?

Comment (koonpeng) [reply packages/dashboard/src/components/tree-button-group.tsx:56]: yea, but we can merge this first and have that in another PR.

**Review by koonpeng — APPROVED** [review]
Thanks for this! Can you make another PR to migrate the components to `react-components`? Try to re-use some of the code also (`SimpleInfoData` for example).

**Comment by matiasbavera** [comment]
> Thanks for this! Can you make another PR to migrate the components to `react-components`? Try to re-use some of the code also (`SimpleInfoData` for example).

Sure thing, leave it to me.
